tls/s2n_fingerprint.c

```diff
@@ -45,278 +34,106 @@ static S2N_RESULT s2n_assert_grease_value(uint16_t val)
-        struct s2n_stuffer *output, struct s2n_hash_state *hash)
-{
-    RESULT_ENSURE_REF(ch);
-
-    bool cipher_found = false;
-    struct s2n_stuffer ciphers = { 0 };
-    RESULT_GUARD_POSIX(s2n_stuffer_init_written(&ciphers, &ch->cipher_suites));
-    while (s2n_stuffer_data_available(&ciphers)) {
-        uint16_t cipher = 0;
-        RESULT_GUARD_POSIX(s2n_stuffer_read_uint16(&ciphers, &cipher));
-        if (s2n_is_grease_value(cipher)) {
-            continue;
-        }
-        RESULT_GUARD(s2n_fingerprint_write_entry(output, &cipher_found, cipher, hash));
-    }
-    return S2N_RESULT_OK;
-}
-
-static S2N_RESULT s2n_fingerprint_write_extensions(struct s2n_client_hello *ch,
-        struct s2n_stuffer *output, struct s2n_hash_state *hash)
-{
-    RESULT_ENSURE_REF(ch);
-
-    /* We have to use the raw extensions instead of the parsed extensions
-     * because s2n-tls both intentionally ignores any unknown extensions
-     * and reorders the extensions when parsing the list.
-     */
-    struct s2n_stuffer extensions = { 0 };
-    RESULT_GUARD_POSIX(s2n_stuffer_init_written(&extensions, &ch->extensions.raw));
-
-    bool extension_found = false;
-    while (s2n_stuffer_data_available(&extensions)) {
-        uint16_t extension = 0, extension_size = 0;
-        RESULT_GUARD_POSIX(s2n_stuffer_read_uint16(&extensions, &extension));
-        RESULT_GUARD_POSIX(s2n_stuffer_read_uint16(&extensions, &extension_size));
-        RESULT_GUARD_POSIX(s2n_stuffer_skip_read(&extensions, extension_size));
-        if (s2n_is_grease_value(extension)) {
-            continue;
-        }
-        RESULT_GUARD(s2n_fingerprint_write_entry(output, &extension_found, extension, hash));
-    }
-    return S2N_RESULT_OK;
-}
-
-static S2N_RESULT s2n_fingerprint_write_elliptic_curves(struct s2n_client_hello *ch,
-        struct s2n_stuffer *output, struct s2n_hash_state *hash)
+S2N_RESULT s2n_fingerprint_hash_add_char(struct s2n_fingerprint_hash *hash, char c)
 {
-    RESULT_ENSURE_REF(ch);
-
-    s2n_parsed_extension *elliptic_curves_extension = NULL;
-    int result = s2n_client_hello_get_parsed_extension(S2N_EXTENSION_SUPPORTED_GROUPS,
-            &ch->extensions, &elliptic_curves_extension);
-    if (result != S2N_SUCCESS) {
-        return S2N_RESULT_OK;
-    }
-
-    struct s2n_stuffer elliptic_curves = { 0 };
-    RESULT_GUARD_POSIX(s2n_stuffer_init_written(&elliptic_curves,
-            &elliptic_curves_extension->extension));
-
-    uint16_t count = 0;
-    RESULT_GUARD_POSIX(s2n_stuffer_read_uint16(&elliptic_curves, &count));
-
-    bool curve_found = false;
-    while (s2n_stuffer_data_available(&elliptic_curves)) {
-        uint16_t curve = 0;
-        RESULT_GUARD_POSIX(s2n_stuffer_read_uint16(&elliptic_curves, &curve));
-        if (s2n_is_grease_value(curve)) {
-            continue;
-        }
-        RESULT_GUARD(s2n_fingerprint_write_entry(output, &curve_found, curve, hash));
+    RESULT_ENSURE_REF(hash);
+    if (hash->hash) {
```

Comment:
Both the JA3-specific and JA4-specific logic will need to branch on whether or not to do the digest, so I wanted something obvious / not based on the internal implementation. I've switched it to an explicit function so that it's clear but there's still only one source of truth.